js.go

```diff
@@ -148,31 +148,33 @@ func (nc *Conn) JetStream(opts ...JSOpt) (JetStream, error) {
 	return js, nil
 }
 
-// JSOpt configures options for the jetstream context.
-type JSOpt func(opts *js) error
+// JSOpt configures a JetStream context.
+type JSOpt interface {
+	configureJSContext(opts *js) error
+}
+
+// jsOptFn configures an option for the JetStream context.
+type jsOptFn func(opts *js) error
+
+func (opt jsOptFn) configureJSContext(opts *js) error {
+	return opt(opts)
+}
 
 func APIPrefix(pre string) JSOpt {
-	return func(js *js) error {
+	return jsOptFn(func(js *js) error {
```

Comment:
Again does this limit it to just us or direct PRs to this codebase?

Reply:
That's right, only direct PRs to this package would be able to create functions that can be applied as an option.

Reply:
As you know, the way we currently the function options, if an option is missing the user can create a function in their own application and would not be blocked waiting on us to provide it (with a new library release). Being OSS, the user can always have a local copy/fork, but with go.mod that becomes a bit more tricky.

So we need to really evaluate the benefit of this approach vs the "risk" it can cause (users blocked and us having to quickly issue a release with the new option). That is a low risk if the library is stable (we have all the options we need), but not sure if that will be the case at this stage with JS.

I like the approach that something becomes generic, such as "max wait", but if we end-up with only that option that benefits from this change, it may not be worth it.

For the record: I am personally not a fan of the "options as functions" approach, I prefer the simple approach of passing a reference to a option struct (and library cloning or making copies of what is needed to not be affected by changes done to the options after the call). But it seems that this is the direction we have taken on several projects, so that's that.

Reply:
Currently all the options for JetStream are private (`pubOpts`, `subOpts`,`*js`), but with the approach in this PR and by making a public options type for example, we actually can have that kind of usage of using a single struct to define all the settings at once.

For example in this [commit](https://github.com/wallyqs/nats.go/commit/8ddd21f183c2df3597957ff43ccdfa26d478a389), I have made a `PubOptions` type that it itself also implements the `PubOpt` interface so it can be used to configure a publish as well, so in that branch all of these would work:

```go
// Publish with default options
js.Publish("baz", msg)

// Publish with custom timeout and expected stream using single options struct.
js.Publish("baz", msg, &nats.PubOptions{MaxWait: 2*time.Second, Stream: "FOO"})

// Publish with custom timeout and expected stream
js.Publish("baz", msg, nats.MaxWait(2*time.Second), nats.ExpectStream("FOO"))
```

In that commit, I made another type `PubOptions` that is basically the same as `pubOptions` which is private, but with longer names and have it as public, but they could be refactored into being a single type later since all the changes are internal to the library.

Also the `PubOpt` interface methods are private in this PR, and defined as:

```go
type PubOpt interface {
	configurePublish(opts *pubOpts) error
}
```

so what users cannot do is to create structs that implement the `PubOpt` interface.  To increase the flexibility, what we can do is instead have the interface be defined in terms of the public type like:

```go
type PubOpt interface {
	configurePublish(opts *PubOptions) error
}
```

Then, if we want users to be able to create their own functional options we can make the following type public in the `nats` package:

```go
// PubOptFn is a function option used to configure a JetStream Publish.
type PubOptFn func(opts *PubOptions) error

func (opt PubOptFn) configurePublish(opts *PubOptions) error {
	return opt(opts)
}
```

With that approach users would be able to create their own functional options  if needed, without having to make `configurePublish` public in the `PubOpt` interface (sample implementation of this approach in this [commit](https://github.com/wallyqs/nats.go/commit/e9e14944baf9397babab96826802cc4e3ff83db2))

```go
	withNameAndTimeout := func(name string, ttl time.Duration) nats.PubOpt {
		return nats.PubOptFn(func(opts *nats.PubOptions) error {
			t.Logf("Configuring my own JS NATS publish...")
			opts.Stream = name
			opts.MaxWait = ttl
			return nil
		})
	}
	if _, err = js.Publish("foo", msg, withNameAndTimeout("TEST", 2*time.Second)); err != nil {
		t.Fatalf("Unexpected publish error: %v", err)
	}
```